src/CodeGen_Vulkan_Dev.cpp

```diff
@@ -2003,18 +2003,25 @@ void CodeGen_Vulkan_Dev::SPIRV_Emitter::visit(const Evaluate *op) {
 }
 
 void CodeGen_Vulkan_Dev::SPIRV_Emitter::visit(const Shuffle *op) {
-    std::cout << " CodeGen_Vulkan_Dev::SPIRV_Emitter::visit(Shuffle): "
-              << "type=" << op->type << " "
-              << "vectors=" << (uint32_t)op->vectors.size() << " "
-              << "is_interleave=" << (op->is_interleave() ? "true" : "false") << " "
-              << "is_extract_element=" << (op->is_extract_element() ? "true" : "false") << "\n";
+    debug(2) << " CodeGen_Vulkan_Dev::SPIRV_Emitter::visit(Shuffle): "
+             << "type=" << op->type << " "
+             << "vectors=" << (uint32_t)op->vectors.size() << " "
+             << "is_interleave=" << (op->is_interleave() ? "true" : "false") << " "
+             << "is_extract_element=" << (op->is_extract_element() ? "true" : "false") << "\n";
+
+    // Sanity check that the total lane count matches between the op-type and indices
+    internal_assert(!op->vectors.empty());
+    for (size_t i = 1; i < op->vectors.size(); i++) {
```

Comment:
I think shuffle nodes are allowed to combine vectors of different sizes. It looks like the removed code handles this, but this assert looks like it would fail.

Reply:
Handling that case looks like it was the whole point of the PR, so I must be missing something. Maybe a longer comment here would help.

Reply:
Yes, good catch!  These asserts seem wrong or out of date ... I lifted them from CodeGen_GPU:

https://github.com/halide/Halide/blame/65b08efa78baf964387b9615b4c80e43d4cb967a/src/CodeGen_GPU_Dev.cpp#L149

And the date on those lines is from 3 years ago.  So my guess is those asserts were written to match the implementation as it was ... without support to handle vectors of different lanes.

Thanks! I'll remove these lines from both CodeGen_Vulkan and CodeGen_GPU.

Reply:
We actually aren't triggering those asserts at all.  Even the test case seems to produce lowered code that has vectors of the same size for the arguments, but only uses a a few of lanes from the arguments to produce the shuffle.  That case wasn't handled correctly before ... it triggered a type mismatch for the assignments which generated invalid code for mixing vector/scalar types of different sizes.